### DISCUSSION TITLE
Venetian blinds tilt (new command ShutterTiltConfig)

### DISCUSSION THREAD

**Discussion by crllsn** [arendst/Tasmota#15974] [Q&A]
Can someone explain in detail what is the logic behind new command ShutterTiltConfig? This is because I'm experiencing some strange behaviour mainly in most closed position and in most open position and I would like to understand why. For example, it looks like someting is wrong when the venetian is fully closed (with tilt closed): I tilt it open (pushing up button) and tasmota thinks it's still fully closed, so pushing down button does nothing. I'm talking about physical button configured as button (not switches). If someone could explain the algorithm in detail I can hopefuly understand if it's not correct in my case or if I'm doing something wrong.
It looks like a control on previous direction should be implemented, but I'm not sure...
Many thanks in advance!!

PS this is how my venetian works:
Imagine it fully closed, tilt closed
- Push UP button: first 1,5s it tilts open, then it starts the up movement.
- If you push DOWN button after a while: first 1,5s it tilts closed and then starts the down movement

**Comment by crllsn** [comment]
I have discovered that something is wrong. I try to explain:
Start condition: venetian closed + tilt closed
If I push the wall button UP and I watch what happens, first the cursor of tilt goes up and after it reaches 100% the position cursor starts moving up. Everything is ok.
Same start condition.
If I push the wall button UP and I watch what happens, the cursor of tilt goes up. If I stop e.g. at 50% of tilt and I press the DOWN wall button, the cursor goes down at 0% BUT nothing actually moves, the relay stays off. 
Any ideas about what could be wrong?

If I do the same thing BUT I just move the tilt UP with wall button (when it's fully closed position)

**Reply by stefanbode** [reply]
Hi, can you please define Shutterbutton1 and 2. and then assign your physical buttons to the defined ones. This may solve all your problems. With physical buttons directly operation the Venetian blind is totally useless. The idea is that after a position change the tilt restores tothetzt previous position automatically. Sometimes there is a small reverse required. But all this makes no sense if you directly switch the relay. Hope this helps

**Reply by crllsn** [reply]
Ok this evening I will try that. If everything goes the right way, I write a couple more instructions to add to venetians section, since (at least to me) this was not obvios at all!
Many thanks for the hint!!

**Reply by crllsn** [reply]
I defined 
```
ShutterButton1 1 up 0
ShutterButton2 2 down 0
```
Let's try.

Blind is closed (position 0, tilt 0)

```
19:45:53.767 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:45:53","Uptime":"10T23:07:28","UptimeSec":947248,"Heap":148,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":52,"Signal":-74,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:45:53.781 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
19:45:53.792 MQT: stat/tasmota_2660C0/POWER2 = OFF
19:45:54.300 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
19:45:54.310 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":0}}
19:46:09.098 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:09","Uptime":"10T23:07:44","UptimeSec":947264,"Heap":147,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"ON","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":3,"Downtime":"0T00:00:33"}}
```
Here I press UP button (less than 2 seconds just to open tilt)
```
19:46:09.113 MQT: stat/tasmota_2660C0/RESULT = {"POWER1":"ON"}
19:46:09.119 MQT: stat/tasmota_2660C0/POWER1 = ON
19:46:09.133 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":1,"Target":100,"Tilt":0}}
19:46:09.809 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":1,"Target":100,"Tilt":42}}
19:46:10.397 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:10","Uptime":"10T23:07:45","UptimeSec":947265,"Heap":148,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":52,"Signal":-74,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:46:10.411 MQT: stat/tasmota_2660C0/RESULT = {"POWER1":"OFF"}
19:46:10.417 MQT: stat/tasmota_2660C0/POWER1 = OFF
19:46:10.928 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
19:46:10.934 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":78}}
```
Ok tilt is now 78 degrees, now I press down button and something strange happens: the blinds do not move (power2 is off), but the tilt goes to -90 (even if my minimum is 0)
```
19:46:16.222 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:16","Uptime":"10T23:07:51","UptimeSec":947271,"Heap":148,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:46:16.239 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
19:46:16.248 MQT: stat/tasmota_2660C0/POWER2 = OFF
19:46:16.282 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:16","Uptime":"10T23:07:51","UptimeSec":947271,"Heap":145,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:46:16.297 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
19:46:16.304 MQT: stat/tasmota_2660C0/POWER2 = OFF
19:46:16.814 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
19:46:16.824 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":-84}}
19:46:19.814 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:19","Uptime":"10T23:07:54","UptimeSec":947274,"Heap":148,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:46:19.828 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
19:46:19.835 MQT: stat/tasmota_2660C0/POWER2 = OFF
19:46:19.849 MQT: tele/tasmota_2660C0/STATE = {"Time":"2022-08-04T19:46:19","Uptime":"10T23:07:54","UptimeSec":947274,"Heap":146,"SleepMode":"Dynamic","Sleep":0,"LoadAvg":999,"MqttCount":4,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Piriweb_domus","BSSId":"46:4E:6D:AF:67:38","Channel":2,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":3,"Downtime":"0T00:00:33"}}
19:46:19.884 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
19:46:19.890 MQT: stat/tasmota_2660C0/POWER2 = OFF
19:46:20.401 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
19:46:20.414 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":-90}}

```
Where should I look? Something is configured wrong but I do not know what...
Some suggestion?

Thanks!

**Comment by stefanbode** [comment]
Ok lets start with the configuration: I assume normal shutter with one relay up and one relay down. Therefore:
```
shutterrelay1 1
interlock 1,2
interlock on
shuttermode 0 (automatic) or 1
shutteropenduration and shuttercloseduration set 
shutterbutton1 and 2 defined 
shuttertiltconfig 0 0 0 0 0
```
restart and test is the shutter move up and down as expected. If this does not work there is something wrong in the general configuration. Now shuttertiltconfig.
Take a stopwatch and try to measure how long does it take for the tilt to go from one extreme into the other extreme. e.g 1.2sec.
a Tasmota timeslice is 0.05sec. Therefore 1.2sec are 24 slices. Your open position is 90° and close 0° (strange. normally at 0° they are horizontal and not closed. May check. The first number must be smaller than the second number. As you can see with relay1 the tilt INCREASE and with relay2 the tilt DECREASE. As long as you not have any reverse defined the position also INCREASE with relay1 and DECREASE with relay2
`ShutterTiltConfig1 0 90 24 90 0`

Typical configurations (germany alu-shutter with tilt)
`ShutterTiltConfig1 -90 90 24 90 0`

Typical Italien Wood venetian
`ShutterTiltConfig1 0 90 24 90 0`

For the first test please do NOT operate the tilt in the END position (closed/open) there are some additional features. I see in the log file that your position==0. Maybe not optimal for a test.

**Reply by crllsn** [reply]
Ok, I'll try a RESET 6 and start from scratch all the blind configuration just to be sure not to have strange old settings.

> Your open position is 90° and close 0° (strange. normally at 0° they are horizontal and not closed. May check. The first number must be smaller than the second number.

Honestly this is strange since my config was

`ShutterTiltConfig1 0 90 40 90 0`

so it's better to start again clean

**Reply by stefanbode** [reply]
Ok, thanks. With this configuration you should never get negative tilts. But I assume there is maybe a side effect/bug because your still in Position==0 when moving back. I have to check this on my testbed.
Your tiltconfig is a valid one. Should work

**Reply by stefanbode** [reply]
This is wrong:
```
ShutterButton1 1 up 0
ShutterButton2 2 down 0
```
Should be:
```
ShutterButton1 1 up 0
ShutterButton1 2 down 0
```
In this case if you're still in position==0 just nothing happens. The tilt do not change. Not ideal but maybe acceptable

**Reply by crllsn** [reply]
Ok, tested and now everything is apparently working as expected! Probably something dirt in the previous configuration...
There's only one thing.
This are the commands I used (first part just general commands):
```
MqttTimeout 15
MqttWifiTimeout 5000
SetOption41 105
ntpserver1 192.168.189.200
ntpserver2 192.168.189.1
PowerRetain 1
SetOption1 1 
SetOption36 3

SetOption80 1
shutterrelay1 1
ShutterMode 1
interlock 1,2
interlock on
shutteropenduration 64.0
shuttercloseduration 64.0
shutterbutton1 1 up 0
shutterbutton2 2 down 0 
shuttertiltconfig 0 90 40 90 0
```

Note that I have buttons on wall and not switches. I understood that 
```
shutterbutton1 1 up 0
shutterbutton2 2 down 0 
```
should give the ability to single, double and hold press.
This is not working: should I use some switchmode to obtain that behaviour?
Even if I have buttons on wall, my template config says GPIOXX SWITCH 1 and GPIOYY SWITCH 2, is that ok or should it be changed?
In any case, try to explain me why since I read but probably not understood correctly.

Please, last effort!-)

**Reply by crllsn** [reply]
did not read your last message sorry! now I correct the shutterbutton options

**Reply by crllsn** [reply]
something strange happens anyway when from fully closed (pos & tilt) I make a quick UP and then I press DOWN: 
```
16:30:33.738 MQT: stat/tasmota_2660C0/POWER1 = ON
16:30:33.748 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":1,"Target":100,"Tilt":0}}
16:30:33.876 MQT: stat/tasmota_2660C0/RESULT = {"POWER1":"OFF"}
16:30:33.884 MQT: stat/tasmota_2660C0/POWER1 = OFF
16:30:34.398 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
16:30:34.404 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":6}}
16:30:36.909 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
16:30:36.921 MQT: stat/tasmota_2660C0/POWER2 = OFF
16:30:36.927 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
16:30:36.937 MQT: stat/tasmota_2660C0/POWER2 = OFF
16:30:37.447 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
16:30:37.455 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":-9}}
16:30:37.646 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
16:30:37.652 MQT: stat/tasmota_2660C0/POWER2 = OFF
16:30:37.660 MQT: stat/tasmota_2660C0/RESULT = {"POWER2":"OFF"}
16:30:37.667 MQT: stat/tasmota_2660C0/POWER2 = OFF
16:30:38.180 MQT: stat/tasmota_2660C0/SHUTTER1 = 0
16:30:38.187 MQT: stat/tasmota_2660C0/RESULT = {"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":-11}}
```

As you can see the UP opens a bit (6) the tilt, then if I press DOWN, POWER2 stays OFF, but tilt goes negative...

when this happens, the thing is messed up and I need to restart.

**Reply by crllsn** [reply]
well this time it was worse, I need to hard reset the module since it crashed and did not recover...

**Reply by stefanbode** [reply]
Are you sure your electrical wire is correct? I have never heard about crash that resets the device.  Ore than strange. Anyhow your configuration is running at my end

**Reply by stefanbode** [reply]
Please assure that you do not stop until the shutter moves a bit. I again see that you stopped the tilt and the position was still 0. this use case might not be supported. Pleas first try it in position 50%

**Reply by crllsn** [reply]
Electrical wiring is ok. My device for your info is a Sonoff dual R3. In pos<>0 everything is ok. The problem happens at pos==0 (very strange that despite I press the button it detects POWER2 = 0). 
I was thinking can it be some sort of ghost switching? Something related to button wired to the unit instead of switches? Perhaps a switchmode 3 could solve this issue?

I think that a workaround should be find. I cannot help since I do not know the logic behind (or if you can show me where to read it, I can study a bit if I can help...).

Can you help me in double press detection too?

**Reply by stefanbode** [reply]
Let me check with if I find a code minimized solution for the position =0. should be the same at position =100. normally if you press up or down I do not change the target tilt. That means that the tilt can restore to the original value if you stop. Never in mind you played with the tilt in the end position. Maybe to if clauses can solve the issue. Let me check. For doublepress you need a button. Not a switch. You can define an additional action like goto 50% on doublepress. This is the most common use case. Tilt can not be changed with doublepress actions. Only positions..

**Reply by crllsn** [reply]
Good to know you try to fix!
BTW I have buttons not switches!-)

For your info: I ask this because most of the time in summer my venetians stay in closed position and I just tilt them to set the amount of light/sun

**Reply by stefanbode** [reply]
I submitted a change to fix the bug. #16174. Please wait until this is merged into DEVELOPMENT version. On my desk this helps with your use-case. The fix is to detect the end-position and if the target is the same to change the target tilt to the correct value. Then everything should move in the right direction

**Reply by crllsn** [reply]
first of all: thanks for your time!

Then I'll test it asap!-)

The multipress is still a mistery though...

**Reply by crllsn** [reply]
I think the problem with multipress is that: I have buttons on wall, but in the template config I have switches assignet to gpio.

Instructions say:
```

ShutterButton<x> <button> <func> <mqtt>
 
This assigns a Tasmota button <button> to control your shutter <x> having functionality <func>. The Tasmota button <button> must already be configured in the module configuration.
```

So I think I have to configure buttons in template configuration. I'll test that!
